src/sb16.js

```diff
@@ -0,0 +1,1607 @@
+    cpu.io.register_read(0x22A, this, this.port2xA_read);
+    cpu.io.register_read(0x22B, this, this.port2xB_read);
+    cpu.io.register_read(0x22C, this, this.port2xC_read);
+    cpu.io.register_read(0x22D, this, this.port2xD_read);
+    cpu.io.register_read(0x22E, this, this.port2xE_read);
+    cpu.io.register_read(0x22F, this, this.port2xF_read);
+
+    cpu.io.register_write(0x220, this, this.port2x0_write);
+    cpu.io.register_write(0x221, this, this.port2x1_write);
+    cpu.io.register_write(0x222, this, this.port2x2_write);
+    cpu.io.register_write(0x223, this, this.port2x3_write);
+    cpu.io.register_write(0x224, this, this.port2x4_write);
+    cpu.io.register_write(0x225, this, this.port2x5_write);
+    cpu.io.register_write(0x226, this, this.port2x6_write);
+    cpu.io.register_write(0x227, this, this.port2x7_write);
+    cpu.io.register_write(0x228, this, this.port2x8_write);
+    cpu.io.register_write(0x229, this, this.port2x9_write);
+    cpu.io.register_write(0x22A, this, this.port2xA_write);
+    cpu.io.register_write(0x22B, this, this.port2xB_write);
+    cpu.io.register_write(0x22C, this, this.port2xC_write);
+    cpu.io.register_write(0x22D, this, this.port2xD_write);
+    cpu.io.register_write(0x22E, this, this.port2xE_write);
+    cpu.io.register_write(0x22F, this, this.port2xF_write);
+
+    cpu.io.register_read(0x330, this, this.port3x0_read);
+    cpu.io.register_read(0x331, this, this.port3x1_read);
+    cpu.io.register_write(0x330, this, this.port3x0_write);
+    cpu.io.register_write(0x331, this, this.port3x1_write);
+
+    bus.register("speaker-tell-samplerate", function(rate)
+    {
+        this.audio_samplerate = rate;
+    }, this);
+
+    bus.register("speaker-process", function(event)
+    {
+        this.audio_process(event);
+    }, this);
+
+    this.reset_dsp();
+}
+
+
+
+//
+// General
+//
+
+
+
+SB16.prototype.reset_dsp = function()
+{
+    this.write_buffer.clear();
+    this.read_buffer.clear();
+
+    this.command = DSP_NO_COMMAND;
+    this.command_size = 0;
+
+    this.dummy_speaker_enabled = false;
+    this.test_register = 0;
+
+    this.dsp_highspeed = false;
+    this.dsp_stereo = false;
+    this.dsp_16bit = false;
+    this.dsp_signed = false;
+
+    this.dac_buffer.clear();
+    this.dac_rate_ratio = 2;
+
+    this.dma_sample_count = 0;
+    this.dma_bytes_count = 0;
+    this.dma_bytes_left = 0;
+    this.dma_bytes_block = 0;
+    this.dma_irq = 0;
+    this.dma_channel = 0;
+    this.dma_autoinit = false;
+    this.dma_buffer_uint8.fill(0);
+
+    this.e2_value = 0xAA;
+    this.e2_count = 0;
+
+    this.sampling_rate = 22050;
+    this.bytes_per_sample = 1;
+
+    this.lower_irq(SB_IRQ_8BIT);
+    this.irq_triggered.fill(0);
+
+    this.asp_registers.fill(0);
+    this.asp_registers[5] = 0x01;
+    this.asp_registers[9] = 0xF8;
+}
+
+SB16.prototype.get_state = function()
+{
+    var state = [];
+
+    // state[ 0] = this.write_buffer              ;
```

Comment:
Please no spaces before the semicolon :-)

Reply:
:astonished: